// @remote-kakao/core/dist/message.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const crypto_1 = require("crypto");
class Message {
    socket;
    sessionEmitter;
    remoteInfo;
    room;
    content;
    sender;
    isGroupChat;
    constructor(data, sessionEmitter, socket, remoteInfo) {
        this.socket = socket;
        this.sessionEmitter = sessionEmitter;
        this.remoteInfo = remoteInfo;
        this.room = data.room;
        this.content = data.content;
        this.sender = {
            name: data.sender,
            getProfileImage: () => data.profileImage,
        };
        this.isGroupChat = data.isGroupChat;
    }
    async reply(text, room = this.room, address = this.remoteInfo.address, port = this.remoteInfo.port, timeout = 60) {
        return new Promise((res, rej) => {
            const session = (0, crypto_1.randomUUID)();
            const data = encodeURIComponent(JSON.stringify({ event: 'sendText', data: { room, text }, session }));
            const handler = (success, data) => {
                if (!success)
                    rej();
                else
                    res(data === undefined ? { success } : { success, data });
                this.sessionEmitter.off(session, handler);
            };
            this.sessionEmitter.on(session, handler);
            this.socket.send(data, 0, data.length, port, address);
            setTimeout(() => rej(), 1000 * timeout);
        });
    }
}
exports.default = Message;
